File: Sliding Window/3. Longest Repeating Character Replacement/solution.rb
def character_replacement(s, k)
  char_counts = Hash.new(0)
  longest_string, left = 0, 0

  s.chars.each_index do |i|
    char_counts[s[i]] += 1
    sliding_window_size = i - left + 1
    character_change_cap = char_counts.values.max + k

    if sliding_window_size > character_change_cap
      char_counts[s[left]] -= 1
      left += 1
    end

    longest_string = [longest_string, i-left+1].max
  end

  longest_string
end

File: Sliding Window/3. Longest Repeating Character Replacement/solution.scala
def characterReplacement(s: String, k: Int): Int = {
  var result = 0
  var count = Array.fill(26)(0)
  var maxCharCountWithinWindow = 0
  var start = 0
  var end = 0
  for (end <- 0 until s.length) {
    // increment the count of curr element in window
    count(s(end) - 'A') += 1
    // get the max repeating char count in window
    maxCharCountWithinWindow = maxCharCountWithinWindow.max(count(s(end) - 'A'))
    // char count of other character apart from max occuring
    var otherCharCount = end - start + 1 - maxCharCountWithinWindow
    // As we can replace till k times only, so till we have other chars count > k, shrink window
    while (otherCharCount > k) {
      count(s(start) - 'A') -= 1
      start += 1
      otherCharCount = end - start + 1 - maxCharCountWithinWindow
    }
    result = result.max(end - start + 1)
  }
  result
}
